### ISSUE TITLE
[BUG] gridsearch is underperforming for-loop

### ISSUE THREAD

**Issue by flockonus** [unit8co/darts#2227]
**Describe the bug**
Learned about grid search feature for hyperparam from this intro video (at the timestamp): https://youtu.be/g6OXDnXEtFA?si=U57Hh3dhuaVvT13b&t=810

To my surprise the official [Quickstart tutorial](https://unit8co.github.io/darts/quickstart/00-quickstart.html) doesn't mention it, but rather does more manual [np.linspace search in a for loop](https://unit8co.github.io/darts/quickstart/00-quickstart.html#Searching-for-hyper-parameters-with-the-Theta-method) The result i got matches the tutorial at:

> The MAPE is: 4.40, with theta = -3.5102040816326543.

But i got curious to do gridsearch for a maybe even better, but in fact it ended up worse no matter what i do 🤔 which leaves me thinking there is a bug somewhere, or at least i can't conciliate the results myself.

This is my snippet, note it takes much much longer to run vs. the above for loop

```python

from darts.utils.utils import SeasonalityMode

parameters = {
    "theta": 2 - np.linspace(-10, 10, 50),
    'season_mode': [SeasonalityMode.MULTIPLICATIVE,SeasonalityMode.ADDITIVE]
}

best_model, best_parameters, score = Theta.gridsearch(
    parameters=parameters,
    series=train,
    start=0.5,
    forecast_horizon=12
)

best_model, best_parameters, score
```

> (Theta(theta=-2.6938775510204085, seasonality_period=None, season_mode=SeasonalityMode.MULTIPLICATIVE),
 {'theta': -2.6938775510204085,
  'season_mode': <SeasonalityMode.MULTIPLICATIVE: 'multiplicative'>},

> 8.051603610304335)  ---- assuming this is the score, worse than 5

**To Reproduce**
Run the tutorial until the point noted above.
OR gist: https://gist.github.com/flockonus/ed2a0149c91d6aaeabab1d74b7ee00ca

**Expected behavior**
I'd expect `Theta.gridsearch` to at least match the manual for-loop results, hopefully better.

**System (please complete the following information):**
 - Python version: 3.10
 - darts version: u8darts-all-0.27.2-pyhd8ed1ab_0 

**Additional context**
I'm new here, possible there is a mistake idk.

**Comment by madtoinou** [comment]
Hi @flockonus,

The hyper-parameters tuning is not covered in the quickstart but has a dedicated page in the [User Guide](https://unit8co.github.io/darts/userguide/hyperparameter_optimization.html) as well as an [example notebook](https://unit8co.github.io/darts/examples/17-hyperparameter-optimization.html).

You were not using the same dataset as the video: your snippet uses `AirPassengers` whereas the video uses `MonthlyMilkDataset`. This alone explains the difference in results.

Furthermore, the values of the `theta` parameters that you tried are rather extreme : the parameters range in the video goes from 0.5 to 2.5 while yours goes for -8 to 12. Since `theta` corresponds to a weighting of the values in this context, it is reasonable to assume that a more optimal parameters is around to the initial best (theta=3 in the video).

Finally, the `score` returned by `gridsearch()` correspond to the `metric` between the forecast of the model trained on the first half of the `series` parameter (since `start=0.5`)  and the second half of this series. In the video, the reported MAPE score is computed after retraining the best model on the entire training series and forecasting the validation series.

The code snippet to reproduce the video results:

```python
import matplotlib.pyplot as plt

from darts import TimeSeries
from darts.models import Theta
from darts.metrics import mape
from darts.utils.utils import SeasonalityMode
from darts.datasets import MonthlyMilkDataset

series = MonthlyMilkDataset().load()
train, test = series.split_after(0.8)

model = Theta()
model.fit(train)
pred = model.predict(len(test))

mape(val, test) # 1.276

# create an internal train/val split
best_model, best_parameters, score = Theta.gridsearch(
    parameters={
        "theta":[0.5, 1, 1.5, 2, 2.5, 3],
        "season_mode": [SeasonalityMode.MULTIPLICATIVE,SeasonalityMode.ADDITIVE]
    },
    series=series1,
    start=0.5,
    forecast_horizon=12,
)

best_model.fit(train)
pred_best = best_model.predict(len(test))
mape(test, pred_best) # 0.978
```

This dataset is a bit tricky because the trend visible in the first 3-thirds of the dataset changes for the last third, making the hyper-parameter tuning a bit challenging as after the training/validation split, the test set does not follow the same distribution as the validation set.

**Comment by flockonus** [comment]
Hi @madtoinou perhaps there was a misunderstanding in the interpretation of how i compared results (it's all `AirPassengersDataset`), i was not trying to compare the video results with the quickstart guide, rather only the quickstart guide. Apologies for the hurried post before.

Also, thank you for the documentation links, I haven't seen another team anywhere near the quality of Unit8 / Darts.

Updated the gist with a hopefully more straightforward comparison here - https://gist.github.com/flockonus/8506f3b784d679c9aedb9f71331d10f5

Output for split of 85%:

```
using split %:  0.85
The MAPE from searching from 50 score: 3.97%, with theta = -0.6530612244897966.

Theta.gridsearch results:
Best parameters:  {'theta': -1.4693877551020407, 'season_mode': <SeasonalityMode.MULTIPLICATIVE: 'multiplicative'>}
Best score: 7.17
```

In this case Theta.gridsearch still finds a different optimal Theta, and the score is worse, but i my high level understanding is that the methods they are being calculated still differ, even tho both are MAPE?

**Comment by madtoinou** [comment]
Ah my bad, I went through the first few lines too quickly. Glad that you find the documentation useful!

In order to compare the outcome of your for-loop with the outcome of `gridsearch()`, you cannot take the returned score as is as it considers several training expanding windows/historical forecasts (called in the background). You would need to add these three lines:

```python
best_model.fit(train)
pred_best_theta_gridsearch = best_model.predict(len(val))
score = mape(val, pred_best_theta_gridsearch)) # 6.27, greater than the "loop model"
```

The parameters returned by `gridsearch()` are not optimal for a particular split as it tries to optimize the "historical" performance of the model on the provided dataset. Meaning that a model is trained on `series` (until the `start` index), predict `forecast_horizon` values and compute the metric. Then, training set is increased by `stride` and the model is fitted gain, ... Once all the possible horizons have been forecasted, the metric is aggregated with the `reduction` argument and the parameters of the model that minimize it are returned. Hence, the parameters returned are supposed to offer the best stability "over time" rather than the best score on a specific split.

You can verify it:

```python
loop_best_model = Theta(best_theta)
loop_best_model_backtest = loop_best_model.backtest(
    series=series,
    start=SPLIT_PCT,
    forecast_horizon=12,
    metric=mape
)

grid_best_model = Theta(**best_parameters)
grid_best_model_backtest = grid_best_model.backtest(
    series=series,
    start=SPLIT_PCT,
    forecast_horizon=12,
    metric=mape
)

assert loop_best_model_backtest > grid_best_model_backtest
```

**Comment by flockonus** [comment]
>  you cannot take the returned score as is as it considers several training expanding windows/historical forecasts (called in the background)

Ah that's the key (and vocabulary) I was missing, thanks for pointing out.

The snippet makes sense as well, thanks for the comprehensive answer.

```
loop_best_model_backtest 7.93
grid_best_model_backtest 7.17
```

Btw, and unrelated as an idea, if there is a training or certification your organization or course were to offer i'd be very excited to take it (online). Unit8 / Darts seems to be off the charts compared to other similar packages like autogluon / automl.
